oyeb/beepbop
# Rewrite macros, add tests and goodies

`State` -> `Context`
--------------------

Renamed `State` to `Context`
* Context now has `:state`, this is the (extra) data. You can notice, `:context`
was renamed to `:state`.
* State struct expanded: `struct` and `valid?`

Tests
-----
* Added tests for all compile-time checks
* Add some tests for `Context`
* Add an Ecto.Repo just for tests
* Added some dummy state machines in `test/example` which are used in
tests.

Miscellaneous
-------------
* You cannot `use Ecto.Schema` and `use BeepBop` in the same module!
* Set up travis builds with CodeCov
* add sane git-hooks
* minor docs improved

Macros (huge rewrite!)
----------------------
* Moved all compile-time checks to `BeepBop.Utils`
* Compute transitions outside `quote`, and passed it inside via
`Module.put_attribute`
* Refactor event macro, break it down
  * The macro was far too complex.
  * Moved the control flow logic to function clauses/pattern matches
* Cleaned up macros, made them... compact.
  - Took inspiration from `Ecto.Schema` and `Plug`

New compile-time checks/features
--------------------------------
* Add check to validate schema is actually a struct
* Also check if the ecto schema has the declared column in it.
  - This requires that the schema module be _already compiled_ before
    this one.
* Added code to handle the case when Schema module is aliased as
_something else_.
* Add `__beepbop__/1` to make metadata accessible
